core/src/main/java/org/apache/iceberg/hadoop/HadoopCatalog.java

```diff
@@ -195,8 +203,108 @@ public void renameTable(TableIdentifier from, TableIdentifier to) {
     throw new UnsupportedOperationException("Cannot rename Hadoop tables");
   }
 
+  @Override
+  public boolean createNamespace(Namespace namespace) {
+    Preconditions.checkArgument(!namespace.isEmpty(), "Your Namespace must be not empty!");
+    Joiner slash = Joiner.on("/");
+    Path nsPath = new Path(slash.join(warehouseLocation, slash.join(namespace.levels())));
+    return io().mkdir(nsPath.toString());
+  }
+
+  @Override
+  public List<Namespace> listNamespaces() {
+    return listNamespaces(Namespace.empty());
+  }
+
+  @Override
+  public List<Namespace> listNamespaces(Namespace namespace) {
+ //   Preconditions.checkArgument(!namespace.isEmpty(), "Your Namespace must be not empty!");
+    List<Namespace> namespaces = new ArrayList<>();
+    String[] nsp;
+    Joiner slash = Joiner.on("/");
+    Path nsPath = new Path(slash.join(warehouseLocation, slash.join(namespace.levels())));
+    FileSystem fs = Util.getFs(nsPath, conf);
+    try {
+      if (!fs.exists(nsPath) || !fs.isDirectory(nsPath)) {
+        throw new NotFoundException("Unknown namespace " + namespace);
+      }
+
+      for (FileStatus s : fs.listStatus(nsPath)) {
```

Comment:
This would be much more concise and easier to read by using either the streams API or Guava `Iterables`:

```java
  Iterable<String> namespaces = Iterables.transform(
      Iterables.filter(
          Iterables.transform(fs.listStatus(nsPath), stat -> stat.getPath()),
          path -> fs.isDirectory(path) && !isTable(fs, path)),
      path -> path.getName());
```

Reply:
thanks Ryan , now I use the stream API rewrite it.